File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    const connection = Array.from(new Array(n + 1), () => []);
    let result = Infinity;
    
    for (let [s, e] of wires) {
        connection[s].push(e);
        connection[e].push(s);
    }
    
    for (let cutting of wires) {
        const visited = new Set();
        
        getTreeDepth(1, visited, cutting);
        
        const diff = Math.abs((n - visited.size) - visited.size);
        result = Math.min(diff, result);
    }
    
    return result;    
    
    function getTreeDepth(current, visited, cutting) {
        visited.add(current);
        
        for (let next of connection[current]) {
            if (current === cutting[0] && next === cutting[1]) continue;
            if (current === cutting[1] && next === cutting[0]) continue;
            if (visited.has(next)) continue;
            
            getTreeDepth(next, visited, cutting);
        }
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
import math

def solution(n, wires):    
    min_diff = 100
    
    for i in range(len(wires)):
        W = [[] for _ in range(n)]
        
        for j in range(len(wires)):
            if i == j:
                continue
                
            s, e = map(int, wires[j])
            W[s-1].append(e-1)
            W[e-1].append(s-1)
            
        stack = [0]
        visited = set([0])

        while len(stack) > 0:
            cn = stack.pop()

            for nn in W[cn]:
                if nn in visited:
                    continue

                visited.add(nn)
                stack.append(nn)
                    
        lv = len(visited)
        min_diff = min(min_diff, abs(n-(2*lv)))
            
    return min_diff